File: stm32-uart/python/SysDebugging/SingleWireTransfering.py
#  This file is part of the DSP-Crowd project
#  https://www.dsp-crowd.com
#
#  Author(s):
#      - Johannes Natter, office@dsp-crowd.com
#
#  File created on 30.11.2019
#
#  Copyright (C) 2019, Johannes Natter
#
#  Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to deal
#  in the Software without restriction, including without limitation the rights
#  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#  copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
#  The above copyright notice and this permission notice shall be included in all
#  copies or substantial portions of the Software.
#
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#  SOFTWARE.

import aEnv

import os
import termios
import tty
import fcntl
import struct
import queue
import socket

from Processing import *

FlowMasterSlave = 0xF0
FlowSlaveMaster = 0xF1

ContentOutCmd = 0xC0

ContentInCmd = 0xC1
ContentInNone = 0x00

DataEnd = 0x00 # String termination
DataCut = 0x17 # End of transmittion block

class SingleWireTransfering(Processing):

	def initialize(self):

		if aEnv.dbgChannel == 'socket':
			self.firstRcvState = self.ContentByteRcv
		else:
			self.firstRcvState = self.FlowControlByteRcv

		# Internal
		self.logLevel = 0
		self.msStart = 0
		self.msLastReceived = 0
		self.frameDone = 0
		self.cmdIdOld = 0
		self.stateSend = self.DbgIfInit
		self.stateRcv = self.firstRcvState
		self.fragments = {}
		self.readBlockSize = 4096

		aEnv.dataIn["cmd"] = {}
		aEnv.dataIn["cmd"]["id"] = 0
		aEnv.dataIn["cmd"]["name"] = ""

		# Environment
		aEnv.devOnline = False

		if aEnv.dbgChannel == 'socket':
			return self.socketCreate()
		else:
			return self.ttyCreate()

	def socketCreate(self):
		self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		self.sock.connect(('localhost', aEnv.dbgPort))
		self.sock.setblocking(False)

		return Positive

	def ttyCreate(self):
		self.TIOCM_RTS_str = struct.pack('I', termios.TIOCM_RTS)
		self.fd = os.open("/dev/ttyUSB0", os.O_RDWR | os.O_NOCTTY | os.O_NONBLOCK)

		old = termios.tcgetattr(self.fd)
		new = old

		# http://man7.org/linux/man-pages/man3/termios.3.html
		new[tty.IFLAG] &= ~termios.ICRNL

		new[tty.OFLAG] &= ~termios.ONLCR

		new[tty.LFLAG] &= ~termios.ECHO
		new[tty.LFLAG] &= ~termios.ICANON

		new[tty.CFLAG] &= ~termios.CBAUD
		new[tty.CFLAG] |= termios.CBAUDEX

		new[tty.ISPEED] = termios.B115200
		new[tty.OSPEED] = termios.B115200

		try:
			termios.tcsetattr(self.fd, termios.TCSANOW, new)
		except:
			termios.tcsetattr(self.fd, termios.TCSANOW, old)
			return self.procErrLog(-1, "tcsetattr() failed")

		return Positive

	def process(self):

		self.stateSend()

		try:
			if aEnv.dbgChannel == 'socket':
				data = self.sock.recv(self.readBlockSize)
			else:
				data = os.read(self.fd, self.readBlockSize)
		except BlockingIOError:
			return Pending

		#if len(data) > 2:
		#	self.procDbgLog("RxD: Received data. len = %d" % len(data))

		for d in data:
			#self.procDbgLog("RxD: Received data: 0x%02X %c" % (d, chr(d)))
			self.data = d
			self.stateRcv()

		return Pending

	# SEND STATES
	def DbgIfInit(self):

		self.procDbgLog("TxD: Initializing debug interface")
		self.cmdSend("aaaaa")

		self.stateSend = self.NextFlowDetermine

	def NextFlowDetermine(self):

		if "cmd" in aEnv.dataOut and aEnv.dataOut["cmd"]["id"] != self.cmdIdOld:
			self.procDbgLog("TxD: Command received from peer")

			self.cmdSend(aEnv.dataOut["cmd"]["name"])
			self.cmdIdOld = aEnv.dataOut["cmd"]["id"]
		else:
			self.dataSend(bytes([FlowSlaveMaster]))

			self.msLastReceived = time_ns() // 10**6
			self.frameDone = 0

			#self.procDbgLog("TxD: Waiting for response")
			self.stateSend = self.ResponseWait

	def ResponseWait(self):

		msCurTime = time_ns() // 10**6

		if msCurTime - self.msLastReceived > 500:
			self.procDbgLog("TxD: Timeout reached for single wire transfer")

			aEnv.devOnline = False
			self.stateRcv = self.firstRcvState

			self.procDbgLog("TxD: Waiting for re-init")
			self.msStart = msCurTime

			self.stateSend = self.ReInitWait
			return

		if not self.frameDone:
			return

		aEnv.devOnline = True
		self.stateSend = self.NextFlowDetermine

	def ReInitWait(self):

		msCurTime = time_ns() // 10**6

		if msCurTime - self.msStart < 1500:
			return

		self.procDbgLog("TxD: Waiting for re-init: done")

		self.stateSend = self.DbgIfInit

	def cmdSend(self, cmd):

		try:
			self.dataSend(bytes([FlowMasterSlave]))
			self.dataSend(bytes([ContentOutCmd]))
			self.dataSend(str.encode(cmd))
			self.dataSend(b"\x00")
		except Exception as e:
			self.procDbgLog(repr(e))

	def dataSend(self, d):
		if aEnv.dbgChannel == 'socket':
			self.sock.send(d)
		else:
			os.write(self.fd, d)
	# END SEND STATES

	# RCV STATES
	def FlowControlByteRcv(self):

		if self.data == FlowMasterSlave:
			self.stateRcv = self.DataIgnore
			self.procDbgLog("RxD: Ignoring data")
		elif self.data == FlowSlaveMaster:
			self.stateRcv = self.ContentByteRcv
			#self.procDbgLog("RxD: Receiving data")

	def DataIgnore(self):

		if self.data == DataEnd or self.data == DataCut:
			self.stateRcv = self.firstRcvState
			self.procDbgLog("RxD: Ignoring data: done")

	def ContentByteRcv(self):

		self.msLastReceived = time_ns() // 10**6

		self.contentId = self.data

		if self.data == ContentInNone:
			self.frameDone = 1
			self.stateRcv = self.firstRcvState
			#self.procDbgLog("RxD: Receiving data: done")
		else:
			if self.contentId != 0xC2:
				self.procDbgLog("RxD: Received content ID from device: 0x%02X" % self.contentId)

			self.stateRcv = self.DataRcv

	def DataRcv(self):

		self.msLastReceived = time_ns() // 10**6

		if self.data == DataEnd:
			if not self.contentId in self.fragments:
				self.procDbgLog("RxD: Got empty message")

				self.stateRcv = self.firstRcvState
				return

			if not self.contentId in aEnv.dataIn:
				aEnv.dataIn[self.contentId] = queue.Queue()

			resp = self.fragments[self.contentId]

			aEnv.dataIn[self.contentId].put(resp)
			del self.fragments[self.contentId]

			if self.contentId == ContentInCmd:
				resp = aEnv.dataIn[self.contentId].get()

				self.procDbgLog("RxD: %s" % resp)

				if self.cmdIdOld != aEnv.dataIn["cmd"]["id"]:
					aEnv.dataIn["cmd"]["id"] = self.cmdIdOld
					aEnv.dataIn["cmd"]["resp"] = resp
				else:
					self.procDbgLog("RxD: Ignored received command response")

			if self.contentId != 0xC2:
				self.procDbgLog("RxD: Receiving data: done")

			self.frameDone = 1
			self.stateRcv = self.firstRcvState
			return

		if self.data == DataCut:
			self.procDbgLog("RxD: Receiving data fragment: done")

			self.frameDone = 1
			self.stateRcv = self.firstRcvState
			return

		if not self.contentId in self.fragments:
			self.fragments[self.contentId] = ""

		self.fragments[self.contentId] += chr(self.data)
		#self.procDbgLog("RxD: Fragment %s" % self.fragments[self.contentId])
	# END RCV STATES

File: targets/stm32/SingleWireTransfering.cpp
/*
  This file is part of the DSP-Crowd project
  https://www.dsp-crowd.com

  Author(s):
      - Johannes Natter, office@dsp-crowd.com

  File created on 30.11.2019

  Copyright (C) 2019, Johannes Natter

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.
*/

#include <string.h>

#include "SingleWireTransfering.h"

#define dForEach_ProcState(gen) \
		gen(StStart) \
		gen(StFlowControlRcvdWait) \
		gen(StContentIdOutSend) \
		gen(StContentIdOutSentWait) \
		gen(StDataSend) \
		gen(StDataSentWait) \
		gen(StContentIdInRcvdWait) \
		gen(StCmdRcvdWait) \

#define dGenProcStateEnum(s) s,
dProcessStateEnum(ProcState);

#if 0
#define dGenProcStateString(s) #s,
dProcessStateStr(ProcState);
#endif

using namespace std;

static char bufRx[2];
static uint8_t bufRxIdxIrq = 0; // used by IRQ only
static uint8_t bufRxIdxWritten = 0; // set by IRQ, cleared by main
static uint8_t bufTxPending = 0; // set by main, cleared by IRQ

uint8_t SingleWireTransfering::idStarted = 0;

enum SwtFlowDirection
{
	FlowCtrlToTarget = 0x0B,
	FlowTargetToCtrl = 0x0C
};

enum SwtContentIdIn
{
	ContentInCmd = 0x1A,
};

enum SwtContentId
{
	ContentNone = 0x15,
	ContentProc = 0x11,
	ContentLog,
	ContentCmd,
};

enum SwtContentEnd
{
	ContentCut = 0x0F,
	ContentEnd = 0x17,
};

SingleWireTransfering::SingleWireTransfering()
	: Processing("SingleWireTransfering")
	, mModeDebug(0)
	, mSendReady(false)
	, mValidBuf(0)
	, mpSend(NULL)
	, mpUser(NULL)
	, mContentTx(ContentNone)
	, mValidIdTx(0)
	, mpDataTx(NULL)
	, mIdxRx(0)
	, mLenSend(0)
{
	mState = StStart;

	mBufInCmd[0] = 0;
	mBufOutProc[0] = 0;
	mBufOutLog[0] = 0;
	mBufOutCmd[0] = 0;
}

/* member functions */

void SingleWireTransfering::fctDataSendSet(FuncDataSend pFct, void *pUser)
{
	mpSend = pFct;
	mpUser = pUser;
}

void SingleWireTransfering::dataReceived(char *pData, size_t len)
{
	char *pDest = &bufRx[bufRxIdxIrq];

	*pDest = *pData;
	(void)len;

	bufRxIdxWritten = bufRxIdxIrq + 1;
	bufRxIdxIrq ^= 1;
}

void SingleWireTransfering::dataSent()
{
	bufTxPending = 0;
}

Success SingleWireTransfering::process()
{
	char data;

	switch (mState)
	{
	case StStart:

		if (!mpSend)
			return procErrLog(-1, "err");

		if (idStarted & cStartedTrans)
			return procErrLog(-1, "err");

		idStarted |= cStartedTrans;
		mSendReady = true;

		mState = StFlowControlRcvdWait;

		break;
	case StFlowControlRcvdWait:

		if (!byteReceived(&data))
			break;

		if (data == FlowCtrlToTarget)
			mState = StContentIdInRcvdWait;

		if (!mModeDebug)
			break;

		if (data == FlowTargetToCtrl)
			mState = StContentIdOutSend;

		break;
	case StContentIdOutSend:

		if (mValidBuf & cBufValidOutCmd) // highest prio
		{
			mValidIdTx = cBufValidOutCmd;
			mContentTx = ContentCmd;
			mpDataTx = mBufOutCmd;
			mLenSend = sizeof(mBufOutCmd);
		}
		else if (mValidBuf & cBufValidOutLog)
		{
			mValidIdTx = cBufValidOutLog;
			mContentTx = ContentLog;
			mpDataTx = mBufOutLog;
			mLenSend = sizeof(mBufOutLog);
		}
		else if (mValidBuf & cBufValidOutProc) // lowest prio
		{
			mValidIdTx = cBufValidOutProc;
			mContentTx = ContentProc;
			mpDataTx = mBufOutProc;
			mLenSend = sizeof(mBufOutProc);
		}
		else
			mLenSend = 0;

		if (mLenSend < 2)
			mContentTx = ContentNone;

		bufTxPending = 1;
		mpSend(&mContentTx, sizeof(mContentTx), mpUser);

		mState = StContentIdOutSentWait;

		break;
	case StContentIdOutSentWait:

		if (bufTxPending)
			break;

		if (mContentTx == ContentNone)
		{
			mState = StFlowControlRcvdWait;
			break;
		}

		// protect strlen(). Zero byte and 'content end' identifier byte must be stored at least
		mLenSend -= 2;
		mpDataTx[mLenSend] = 0;

		mLenSend = strlen(mpDataTx);

		mpDataTx[mLenSend] = 0;
		++mLenSend;

		mpDataTx[mLenSend] = ContentEnd;
		++mLenSend;

		mState = StDataSend;

		break;
	case StDataSend:

		bufTxPending = 1;
		mpSend(mpDataTx, mLenSend, mpUser);

		mState = StDataSentWait;

		break;
	case StDataSentWait:

		if (bufTxPending)
			break;

		mValidBuf &= ~mValidIdTx;

		mState = StFlowControlRcvdWait;

		break;
	case StContentIdInRcvdWait:

		if (!byteReceived(&data))
			break;

		if (data == ContentInCmd && !(mValidBuf & cBufValidInCmd))
		{
			mIdxRx = 0;
			mBufInCmd[mIdxRx] = 0;

			mState = StCmdRcvdWait;
			break;
		}

		mState = StFlowControlRcvdWait;

		break;
	case StCmdRcvdWait:

		if (!byteReceived(&data))
			break;

		if (data == FlowTargetToCtrl)
		{
			mBufInCmd[0] = 0;
			mState = StContentIdOutSend;
			break;
		}

		if (data == ContentEnd)
		{
			mBufInCmd[mIdxRx] = 0;
			mValidBuf |= cBufValidInCmd;

			mState = StFlowControlRcvdWait;
			break;
		}

		if (mIdxRx >= sizeof(mBufInCmd) - 1)
			break;

		mBufInCmd[mIdxRx] = data;
		++mIdxRx;

		break;
	default:
		break;
	}

	return Pending;
}

uint8_t SingleWireTransfering::byteReceived(char *pData)
{
	uint8_t idxWr = bufRxIdxWritten;

	if (!idxWr)
		return 0;

	--idxWr;
	*pData = bufRx[idxWr];

	bufRxIdxWritten = 0;

	return 1;
}

void SingleWireTransfering::processInfo(char *pBuf, char *pBufEnd)
{
#if 0
	dInfo("State\t\t%s\n", ProcStateString[mState]);
#else
	(void)pBuf;
	(void)pBufEnd;
#endif
}

/* static functions */

